### PR TITLE
[BinFmt] Error on serialization custom events in by default in .NET framework

### LINKED ISSUE #9300: [BinFmt] Custom Build Event Args will lead to build warning/error by default
### Context
https://github.com/dotnet/msbuild/issues/6215

Per:
https://liquid.microsoft.com/Web/Object/Read/MS.Security/Requirements/Microsoft.Security.SystemsADM.10010
https://twcsecurityassurance.visualstudio.com/SecurityPolicy/_git/SecurityPolicy/pullrequest/1230

> Existing uses of BinaryFormatter must be removed or disabled by Jan 1, 2024.

### Expected outcome
* Behavior of .NET and .NET Framework version of MSBuild is unified - the the BinFmt is disallowed by default
* While BinFmt is still workable on Framework - assume it is not for the purpose of this change (as it may be soon removed from Framework and only be pluggable via optional nuget) - so it is fine to drop violating events (providing build error is emitted)

**UPDATE (Oct/19):** 
* For more gradual introduction of this breaking change, let's implement this for .NET Framework as a warning and without discarding the event. 
* .NET Core behavior will stay the same. 
* Per offline discussion with @rokonec - this can be achieved by skipping the sender side check for Framework, as receiving side check already has a warning.
* In the future we'll change the warning to error (not part of this item)

### Open Questions
* Does `AppContext.TryGetSwitch("System.Runtime.Serialization.EnableUnsafeBinaryFormatterSerialization", out bool enabled)` make sense on .NET Framework as well. ~~I'll provide answer to this later on.~~ - the switch can be used on Full Framework as well. There is currently no usage of that in Framework runtime. But we can keep the code for boh versions to avoid special casing

### PR DESCRIPTION
**Pull request by rokonec**
Fixes #9300 

### Context
.NET Framework shall throw error when attempting to serialize customer event args by binfmt.
See #9300 for more context.

### Changes Made
- testing switch in .net framework 4 as well (might revisit)
- error thrown by default

### Testing
Local. Unit tests.

### Notes
It is breaking change, but chance of break is small, IMO.

### CODE REVIEW

**Review by JanKrivanek — APPROVED** [review]
Looks good to me.

The changewave should be 17.9 unless you are explicitly convinced it should stay under 17.8

File: src/Build/BackEnd/Node/OutOfProcNode.cs [hunk 1]
```diff
@@ -584,18 +584,16 @@ private void SendPacket(INodePacket packet)
         {
             if (_nodeEndpoint.LinkStatus == LinkStatus.Active)
             {
-#if RUNTIME_TYPE_NETCORE
                 if (packet is LogMessagePacketBase logMessage
                     && logMessage.EventType == LoggingEventType.CustomEvent 
-                    &&
-                    (ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave17_8) || !Traits.Instance.EscapeHatches.IsBinaryFormatterSerializationAllowed)
+                    && (ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave17_8) || !Traits.Instance.EscapeHatches.IsBinaryFormatterSerializationAllowed)
```
[the remark this answers is not in the document]
Comment (rainersigwald) [reply src/Build/BackEnd/Node/OutOfProcNode.cs:589]: 17.10, actually, per the current policy we have only even changewaves.

**Comment by rainersigwald** [comment]
Can you remind me what partner team buyoff we have on this? Are all known first-party uses rectified?

**Comment by rokonec** [comment]
@rainersigwald I have collected some usage data here:
https://github.com/dotnet/msbuild/issues/8825#issuecomment-1582161894

I think most of those are in-process instantiation so it will not be serialized for IPC and hence this change will not break them.

Now when I think about it more, chance of this change to cause serious havoc is small. I have rephased NOTE

File: src/Build/BackEnd/Node/OutOfProcNode.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (rokonec) [reply src/Build/BackEnd/Node/OutOfProcNode.cs]: done

**Comment by rainersigwald** [comment]
Can you run an experimental insertion and try out the build with C++ code analysis and FxCop? I would expect the custom errors to be created at build time and thus in the MSBuild.exe worker, not in devenv.exe.

File: src/Build/BackEnd/Node/OutOfProcNode.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/Build/BackEnd/Node/OutOfProcNode.cs]: Please mention the changewave in .md documentation

[the remark this answers is not in the document]
Comment (rokonec) [reply src/Build/BackEnd/Node/OutOfProcNode.cs]: What .md doc?

[the remark this answers is not in the document]
Comment (JanKrivanek) [reply src/Build/BackEnd/Node/OutOfProcNode.cs]: > Can it be mentioned in the list of changes?: https://github.com/dotnet/msbuild/blob/main/documentation/wiki/ChangeWaves.md#current-rotation-of-change-waves

[the remark this answers is not in the document]
Comment (rokonec) [reply src/Build/BackEnd/Node/OutOfProcNode.cs]: done

**Comment by JanKrivanek** [comment]
Currently blocked by https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1908682

File: src/Framework/Traits.cs [hunk 2]
```diff
@@ -392,7 +392,9 @@ public bool EnableWarningOnCustomBuildEvent
 #if RUNTIME_TYPE_NETCORE
                     return true;
 #else
-                    return false;
+                    return ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave17_10) ?
+                        true :
+                        false;
```
Comment (ladipro) [inline src/Framework/Traits.cs:397]:
```suggestion
                    return ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave17_10);
```
.. or does it make it less readable/obvious?

**Review by ladipro — APPROVED** [review]

**Pull request merged**